File: content/snippets/typescript/interview_lcci/02/08_linked_list_cycle.ts
class ListNode {
  val: number;
  next: ListNode | null;

  constructor(val: number, next: ListNode | null = null) {
    this.val = val;
    this.next = next;
  }
}

function buildList(nums: number[]): ListNode | null {
  const dummy = new ListNode(0);
  let tail = dummy;
  for (const n of nums) {
    tail.next = new ListNode(n);
    tail = tail.next;
  }
  return dummy.next;
}

function hasCycle(head: ListNode | null): boolean {
  let slow: ListNode | null = head;
  let fast: ListNode | null = head;
  while (fast !== null && fast.next !== null) {
    slow = slow!.next;
    fast = fast.next.next;
    if (slow === fast) {
      return true;
    }
  }
  return false;
}

function main(): void {
  const a = buildList([1, 2, 3, 4]);
  let tail = a;
  while (tail !== null && tail.next !== null) {
    tail = tail.next;
  }
  if (tail !== null && a !== null && a.next !== null) {
    tail.next = a.next;
  }

  const b = buildList([1, 2, 3, 4]);

  console.log(hasCycle(a));
  console.log(hasCycle(b));
}

main();

File: content/snippets/cpp/interview_lcci/02/08_linked_list_cycle.cpp
#include <iostream>
#include <vector>

struct ListNode {
    int val;
    ListNode* next;
    ListNode(int v) : val(v), next(nullptr) {}
};

ListNode* buildList(const std::vector<int>& nums) {
    ListNode dummy(0);
    ListNode* tail = &dummy;
    for (int n : nums) {
        tail->next = new ListNode(n);
        tail = tail->next;
    }
    return dummy.next;
}

bool hasCycle(ListNode* head) {
    ListNode* slow = head;
    ListNode* fast = head;
    while (fast && fast->next) {
        slow = slow->next;
        fast = fast->next->next;
        if (slow == fast) return true;
    }
    return false;
}

int main() {
    ListNode* a = buildList({1, 2, 3, 4});
    ListNode* tail = a;
    while (tail->next) tail = tail->next;
    tail->next = a->next; // create cycle

    ListNode* b = buildList({1, 2, 3, 4});

    std::cout << std::boolalpha << hasCycle(a) << '\n';
    std::cout << std::boolalpha << hasCycle(b) << '\n';
    return 0;
}
